Workspace Manipulation / ADS docs (#28507)
mantidproject/mantid: docs/source/concepts/scripting_workspaces.rst

.. _scripting_workspaces:

=====================
Workspaces in Scripts
=====================

The :ref:`Workspaces Toolbox<WorkbenchWorkspaceToolbox>` is used to store and manage workspaces within Mantid. Behind the scences, the Analysis Data Service or ADS is used to control these workspaces. 

Below are some examples of how to control workspaces with a script.

.. contents:: Table of contents
    :local:

Relevant Algorithms
===================

Required import:

.. code-block:: python

    from mantid.simpleapi import *

:ref:`Load a Workspace <algm-Load>` from a file:

.. code-block:: python

    ws = Load('EMU00020884.nxs')

:ref:`Create a Workspace <algm-CreateWorkspace>`:

.. code-block:: python

    ws = CreateWorkspace(DataX=X, DataY=Y, DataE=E, NSpec=4, UnitX="Wavelength")

:ref:`Delete a Workspace <algm-DeleteWorkspace>`:

.. code-block:: python

    DeleteWorkspace(ws)

:ref:`Group Workspaces <WorkspaceGroup>`:

.. code-block:: python

>     wsGroup = GroupWorkspaces("ws1,ws2,ws3")

Reviewer: Probably better to encourage people to use lists rather than long strings
```suggestion
    ws1 = CreateSampleWorkspace()
    ws2 = CreateSampleWorkspace()
    ws3 = CreateSampleWorkspace()
    wsGroup = GroupWorkspaces([ws1,ws2,ws3])
    
    # or if you only have the names of the workspaces 
    wsGroup = GroupWorkspaces(['ws1','ws2','ws3'])
```

:ref:`Save a Workspace in a supported ASCII format <algm-SaveAscii>`:

.. code-block:: python

>     SaveAscii(InputWorkspace=ws1,Filename=savefile,Separator="CSV")

Reviewer: ```suggestion
    SaveAscii(InputWorkspace=ws,Filename=savefile,Separator="CSV")
```

:ref:`Save a Workspace in Nexus format <algm-SaveNexus>`:

.. code-block:: python

>     SaveNexus(InputWorkspace=ws1,Filename=savefile.nxs)

Reviewer: ```suggestion
    SaveNexus(InputWorkspace=ws,Filename=savefile.nxs)
```

:ref:`Plot Spectra from a Workspace <mantidplot.plotSpectrum>`:

.. code-block:: python

>     plotSpectrum(ws,spectrum_nums=[1,2,3],error_bars=True, waterfall=False)

Reviewer: I think this needs a simple example first without the rest and then expand,  also we probably can't use the spectrum_nums argument as that is workbench only
```suggestion
    #simple plot of 1 spectra
    plotSpectrum(ws, 0)
    
    #Plot with 3 spectra, and error bars
    plotSpectrum(ws, [0,1,2], error_bars=True)
    
    # Waterfall plot from two workspaces, with two spectras from each
    plotSpectrum([ws1, ws2], [0,1], waterfall=True)
    
```

Reviewer: Unless you explain that, you could have this bit be plotting by workspace index.  and then another one about plotting by spectrum number, and mention that this is only supported in the workbench., then you can use the spectrum number argument.

Workspace Management
====================

Required import:

.. code-block:: python

    from mantid.api import AnalysisDataService as ADS
    #to skip this import, use 'mtd' instead of 'ADS'

Access a workspace, loaded in the Workspaces Toolbox, inside a script:

.. code-block:: python

    ws = ADS.retrieve('ws')

Access ALL workspaces,loaded in the Workspaces Toolbox, inside a script:

.. code-block:: python

>     ADS.importAll()

Reviewer: ```suggestion
    # This will create a python variable matching the workspace name for each workspace in the ADS
    ADS.importAll()
```

Get a list of currently loaded workspaces:

.. code-block:: python

    ws_names = ADS.getObjectNames()

Delete all Workspaces:

.. code-block:: python

>     ADS.clear('ws')

Reviewer: doesn't need the argument
```suggestion
    ADS.clear()
```
